### PR TITLE
Compare two case setup

### LINKED ISSUE #2375: All two run tests failing - no namelists in case2
No namelist is produced for case2 of two case runs, causing a runtime error.

### PR DESCRIPTION
**Pull request by jedwards4b**
The PR #2373 causes some two test system test not to populate the case2 run directory with namelists.   This also occurs for the ERI test which clones ref1 and ref2 cases.   Fixed by explicitly running case_setup for both phases of comparetwo type tests and for each clone of the ERI test 

Test suite: cime_developer, 
Test baseline: 
Test namelist changes: 
Test status: bit for bit

Fixes #2375 

User interface changes?: 

Update gh-pages html (Y/N)?:

Code review:

### CODE REVIEW

File: scripts/lib/CIME/SystemTests/system_tests_compare_two.py [hunk 1]
```diff
@@ -231,7 +232,7 @@ def run_phase(self, success_change=False):  # pylint: disable=arguments-differ
             with self._case2:
                 logger.info('Doing second run: ' + self._run_two_description)
                 self._activate_case2()
-                # we need to make sure run2 is properly staged.
+               # we need to make sure run2 is properly staged.
```
Comment (jgfouca) [inline scripts/lib/CIME/SystemTests/system_tests_compare_two.py]: Intentional? Looks like a bad indent.

**Comment by jgfouca** [comment]
We'd be able to delete even more lines if the base class implementation of case_N_setup was "pass" instead of NotImplemented error.

**Comment by billsacks** [comment]
Thanks, @jedwards4b . This basically looks good, but I have one question before I approve: It looks like the previous code wasn't consistent about whether `test_mode` was set in the call to `case_setup`, either for case one or for case two. The new, shared code forces consistency in this respect. Is this going to cause any changes in behavior for any of the tests?

**Comment by billsacks** [comment]
> We'd be able to delete even more lines if the base class implementation of case_N_setup was "pass" instead of NotImplemented error.

I'd object to that change. I think that forcing all of the tests to implement case_N_setup helps ensure that the API is being used correctly. That said, I don't feel super strongly about it, so I'd be open to being overruled.

File: scripts/lib/CIME/SystemTests/system_tests_compare_two.py (hunk 1 above)
Comment (jedwards4b) [reply scripts/lib/CIME/SystemTests/system_tests_compare_two.py]: fixed

**Comment by jedwards4b** [comment]
unit tests are failing - I'll look into that tomorrow.

**Comment by jedwards4b** [comment]
Oh, @billsacks I did note the inconsistent behavior and tested that the choice I made worked in all cases.

**Comment by billsacks** [comment]
Thanks a lot, @jedwards4b !

**Comment by billsacks** [comment]
@jedwards4b I think the problem here is that we can't call `case_setup` on a fake case object. The solution that would make sense to me would be similar to what we did a few months ago for `create_clone`: create a pass-through `case_setup` method in the `Case` class. Then have `case_fake.py` also implement `case_setup`, but there it will simply be implemented as `pass`. Then we'd change calls that look like `case_setup(case, ...)` to `case.case_setup(...)`. I'm not sure if I'm explaining that clearly. I can take a stab at implementing it if you'd like me to.

**Comment by billsacks** [comment]
Looks good, thanks.

**Comment by jedwards4b** [comment]
This change causes me to once again call into question the value of this style of test.   The test failure did nothing to elucidate any new problem in the code and caused me to inject a fake pass through function in the case class which does nothing to further code development.  All tests are now passing.

**Comment by rljacob** [comment]
What style of test do you mean?

**Comment by jedwards4b** [comment]
@rjacob test_unittests (__main__.A_RunUnitTests)

**Comment by jedwards4b** [comment]
Maybe we should consider incorporating some methods like case_clone, case_setup etc into the case class with something like http://www.qtrac.eu/pyclassmulti.html ?   This would remove the need for these wrappers in the case class.

**Comment by billsacks** [comment]
@jedwards4b I'll admit that, whenever you have run into a problem like this, I also question whether I made the right choice in how I set up these unit tests. I keep reaching the tentative conclusion that this is at least a good a choice as the alternative, but I'm not that confident in my own assessment. The alternative (I think) would be to use a real case for each of the unit tests. In addition to being slower (how much slower, I'm not sure, but I suspect significantly slower), I suspect that would come with its own maintenance problems. But I don't have a lot of intuition here.

I'll also add, though, that I personally feel that having the case object control the case_setup behavior is good design: I feel like, as a general rule, a given object should control its own behavior. I have found it unintuitive that case_setup lives entirely outside of the Case class. There's good reason for this, to prevent Case from growing too large, but I personally prefer the design where users of the case class call `case_setup` by calling a method on the case object, rather than by calling some totally separate function that lives outside of Case.

**Comment by billsacks** [comment]
@jedwards4b If I understand http://www.qtrac.eu/pyclassmulti.html correctly, it seems to be advocating for using inheritance for a relationship that I would more intuitively view as composition.

I think there are really two related questions: (1) What is the "best" design for the Case class and various things that operate on the case; (2) Should we use a fake Case class for some unit tests? The following is largely an elaboration of my previous comment.

(1) What is the "best" design for the Case class and various things that operate on the case? I feel that it would be easier for code that uses case objects if there were a consistent rule that, when you're doing something to a case, you do so via method calls on the case object. We currently aren't consistent in this respect (`case_setup` being a case in point), which makes it harder to know where to look when you want to figure out how to do something. I acknowledge that this will grow Case's already-large interface even larger, but I'd argue that this is better than the current state of affairs, which makes it _appear_ like case_setup isn't really part of the interface of case, when in fact (for all intents and purposes) it is. 

I feel like a pretty common object-oriented design technique to deal with large classes like this is to then delegate particular behaviors to other classes. So, from the perspective of a user of a case object, they would only need to know about the interface of Case. But to keep the Case implementation manageable, it would delegate some of its implementation to other classes via pass-through methods. Sure, pass-through methods have a bit of overhead, but I personally see them as providing more benefits than costs in situations like this.

From this perspective, the need to introduce a `case_setup` pass-through method for the sake of the unit tests indicated an issue with the original design, and introducing this method actually improves the design - not just the testability.

(2) Should we use a fake Case class for some unit tests? If, for example, others don't agree with my assessments in (1), then it would be more challenging to use a fake Case class for unit tests. Is that a problem? This could be evaluated based on (a) the performance implications for testing, (b) the extra effort it takes to write unit tests using a real Case object, and (c) the extra effort it takes to maintain unit tests using a real Case object, compared to the effort needed to maintain the fake Case class.

**Pull request merged**